pkg/common-controller/snapshot_controller.go

```diff
@@ -1095,10 +1136,30 @@ func (ctrl *csiSnapshotCommonController) updateSnapshotStatus(snapshot *crdv1.Vo
 	if updated {
 		snapshotClone := snapshotObj.DeepCopy()
 		snapshotClone.Status = newStatus
+
+		// We need to record metrics before updating the status due to a bug causing cache entries after a failed UpdateStatus call.
+		// Must meet the following criteria to emit a successful CreateSnapshot status
+		// 1. Previous status was nil OR Previous status had a nil CreationTime
+		// 2. New status must be non-nil with a non-nil CreationTime
+		driverName := content.Spec.Driver
+		if !utils.IsSnapshotCreated(snapshotObj) && utils.IsSnapshotCreated(snapshotClone) {
+			createOperationKey := metrics.NewOperationKey(metrics.CreateSnapshotOperationName, snapshot.UID)
+			ctrl.metricsManager.RecordMetrics(createOperationKey, metrics.NewSnapshotOperationStatus(metrics.SnapshotStatusTypeSuccess), driverName)
+		}
```

Comment:
Ah I see - yes this would cause a leak. For example, if the CreateSnapshot operation took 1 reconcile to record, and CreateSnapshotAndReady took 5 reconciles to record, we would have 4 leaked CreateSnapshot operations. We can drop the metric in this case if we determine the old and new snapshots are already created.

Reply:
Added a case to handle this.

Reply:
should we do this in "syncUnreadySnapshot" instead of here? i.e., if a snapshot has been cut, do not add the "createOperationKey" operation instead of an explicit dropping. I think that is a more clean solution. WDYT? @ggriffiths

Reply:
Moved this to start in syncUnreadySnapshot only when the snapshot has not been created yet.